docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,121 @@
+To facilitate this separation, it seems necessary to introduce a mechanism to export/import deployed data between Wing apps.
+The rest of this proposal does it via `cloud.Outputs`, but that likely deserves its own RFC.
+
+## Example Situation
+
+Acme Co. has a platform team and several application teams. Their platform team manages Kubernetes clusters in AWS with additional AWS resources as needed.
+Application teams write applications in TypeScript and deploy them to these shared Kubernetes clusters.
+The centralized infrastructure definitions are written in Wing in their own repository.
+The applications are written in TypeScript but also have Wing files for infrastructure/deployment.
+
+To add a new application, the platform team needs to ensure that the application has the necessary permissions to interact with the shared resources. To do this, a kubernetes namespace and service account is created:
+
+```wing
+let namespace = cluster.createNamespace("app123");
+let serviceAccount = cluster.createServiceAccount(
+  namespace: namespace,
+  name: "app123-svc",
+);
+
+serviceAccount.grantRead(sharedBucket);
+```
+
+Applications using that namespace and service account will now have access to the shared bucket.
+
+To export this data to downstream applications, `cloud.Outputs` can be used:
+
+```wing
+let outputs = new cloud.Outputs();
+
+// Can reference preflight data (tokens will be resolved)
+outputs.add("namespace", namespace.name);
+outputs.add("serviceAccount", serviceAccount.name);
+
+// and inflight clients (includes the necessary information to interact with the resource for the given operations)
+// (e.g. a bucket name, a URL, etc.)
+outputs.lift("sharedBucket", sharedBucket, ["put"]);
+
+// Create a Wing library
+outputs.export(
+  name: "@acme/core-infra",
+);
+```
+
+This will create a centralized registry that can be imported elsewhere.
+During deployment, this registry will be updated with the relevant information.
+Additionally, it exports a wing library with all the necessary types and runtime code to look up that information.
+This library is published to an internal registry for other teams.
+
+After the central infrastructure is deployed, the application team can import the information into their own wing application:
+
+```wing
+bring "@acme/core-infra" as coreInfra;
+
+let workload = new Workload(
+  namespace: coreInfra.namespace,
+  serviceAccount: coreInfra.serviceAccount,
+  image: join(@dirname, "app"),
+);
+```
+
+Now the application will have the necessary permissions to interact with the shared bucket.
+
+However, we need to create a client for TypeScript to actually make use of these permissions.
+The environment where the client will run needs to know where to resolve the bucket.
+To do this, the client can attach to any inflight host. This takes care of preparing it for the client to run.
+Luckily, the workload exposes this host for us:
+
+```wing
+let client = new TypeScriptClient(
```

Comment:
Feels like this needs to be defined at the infra wing level and published to be used by the application.

Reply:
How would the client know how to resolve the information?
I know that's basically the point of this RFC, but if the answer is "that's completely up to the user" then this RFC could be shortened to "TypeScriptClient writes types and inflight clients to disk, then the users can figure out how to make use of that"

Reply:
Yes. Ideally this is a simple low level mechanism but it's not enough to just write stuff to disk because some binding context is only resolved during deployment (eg bucket ARN). I agree that one major piece is generating the client and the other is some kind of binding context protocol that is created during deployment exported and then passed to the app when it starts so it can give it to the client.